Parallax for units [LLM assisted]

mentos987 opened the pull request:
Enables parallax visuals for units by using new input: parallax_mult

Used inside unit.cfg files like:
```
[frame]
   image="units/bats/bloodbat-ne-[3,4].png:[1,100]"
   parallax_r=1.05
[/frame]
```

This is mostly useful for flying units.

Further improvements along this route would probably entail separating/segmenting the flying unit from its shadow, and having no parallax on the shadow. We could then place the flying unit further up from its shadow and increase the parallax a bit.

This image separation/segmentation would also allow ground units to have parallax effects (you could segment the legs into one [frame] and the upper body in another and run different parallax on them)

The video here is rather choppy and thus difficult to catch the parallax bats in it (it looks better in-game)

https://github.com/user-attachments/assets/d6b52ae4-bddb-49bb-b3a4-f150f4185b92

mentos987 committed:
Parallax for units

Enables parallax visuals for units by using new input parallax in [frame] tags in unit.cfg files.

This is mostly useful for flying units.

newfrenchy83 commented:
another indentation fix is needed.

doofus-01 commented:
I'm a big fan of enabling parallax scrolling, but I'm not sure how well it applies to units.  I will have to try this PR, it's too hard to tell from the posted examples.  Thanks for the effort in any case, I will not let this review languish.

mentos987 commented:
Remade blood bat as discussed above. Here is the result.


https://github.com/user-attachments/assets/aaddd2ec-3bab-4c6d-8bd3-772e6df283cd

doofus-01 commented:
I've tested it out, and I'm sorry to say I don't see what the intended effect here is.  The most noticeable effect was due to the `[shadow_frame]` animation, so I bumped the parallax_mult to 1.65, and it looks like the effect is to have the light source scroll around faster than the unit, which scrolls faster than the scenery, away from the center of the view?  There is definitely room for more shadow effects, but I'm not seeing parallax scrolling playing a strong role.

I don't mean to be overly critical of the blood bat demo, I realize it is just a quick example.  Maybe there is some other use case that makes sense.  If this becomes one of the common keys for all image-related tags, that's probably a good thing.

If this does get merged, can we revert the blood bat animation?  If needed, it could be a part of a test scenario.

mentos987 commented:
@doofus-01 Strange that you would not see it. I see the parallax clearly with 1.05 as multiplier. 1.65 would cause the image to be offset by several tiles for me. Do you not get the parallax sense of height? That the flying units are at a higher altitude than the terrain? It is supposed to be subtle, but defiantly noticeable.

`If this does get merged, can we revert the blood bat animation?` -- it is only here for testing. Will commit it out later.

doofus-01 commented:
> Could be a matter of us having different screen resolutions or something. Will have to test.

If I know what I'm looking for and I'm really looking for it, I can see it is working.  But I don't think the overall effect is convincing, at least not without some major shadow graphics overhaul.

As I indicated earlier, there might be some other use case (towering tentacle, Eeyore rain-cloud, etc.), so I'm not opposed to this.  But I'm not seeing it as useful on its own.  If other graphics-related tags have a parallax key, and this is in line with that, then it makes sense to include and maybe UMC can do something with it.

mentos987 commented:
Alright, then we can add it as an input option but not use it for any units (for now).

Will probably rename it from "parallax_mult" to "parallax".

mentos987 commented:
I have removed the test unit and cleaned up a bit. 

There are optimisations to be made but they would clash with https://github.com/wesnoth/wesnoth/pull/10609 a lot so I will just leave a few TODO comments in there for now. It is not required.

This is ready for merge.

will-ca commented:
Is nobody going to point out this makes it hard to see which hex the unit is on? For both the bats and the rocs.

I'm not sure this is a good idea unless 2.5D effects are added in a consistent way for everything. For example, do bats fly 10X higher than the height of a mountain, or a castle tower? If not, what's your brain supposed to think when panning makes the bats show visible parallax, but not anything else?

mentos987 commented:
These videos are showcase only, extreme examples of parallax to make it clear what the effect is. Any in-game use will need to be calibrated with the things you mentioned in mind. @will-ca

Pentarctagon commented:
@Dalas121 @Hejnewar @knyghtmare is this something either of you would see yourselves/UMC authors using?

Dalas121 commented:
I don't think it's a high priority feature, but I can think of a few places where it would be nice to have: wind or rain, for example.

Hejnewar commented:
Units should be relatively close to the ground to A) not obscure other units B) to be easy to select.
But for effects it's fine.
